### ISSUE TITLE
Adding new folder/file

### ISSUE THREAD

**Issue by votiakov** [Plaristote/sharepoint-ruby#20]
Hello!

I am trying to create an empty folder in Sharepoint Online from my rails app.
I get an exception with this message: _The security validation for this page is invalid. Click Back in your Web browser, refresh the page, and try your operation again._

Authentication is successful and I can successfully send GET requests, but POST requests don't work. I tried add_folder, add_file and also tried sending requests directly via site.query.

Most probably it is not a sharepoint-ruby issue but a problem with my sharepoint configuration, but I kindly ask you for support if you had this issue or know how to solve it.

Best regards

**Comment by Plaristote** [comment]
It is most likely a sharepoint-ruby issue: I haven't worked on Sharepoint sites for a while, but I'm fairly certain that's the kind of response you get when you send an update/delete query without the proper `etag` attribute.

There's no `etag` attributes on create queries though, so maybe sharepoint-ruby isn't doing exactly the query it's supposed to do. Looking at the code, sounds like both create and update queries are made using the `POST` method... so, there might be some confusion here, maybe it's doing an update query instead of a create one.

Also, I believe sharepoint-ruby has been developed for Sharepoint 2013 (I started this a while ago). So, if you're using Sharepoint 2016, maybe we have a few things to tweak before it works with that new version.

Anyhow, this might not be easy to fix.

You can set the `verbose` attribute of your `Sharepoint::Site` object to true, to have more details on your query. It might be useful to compare your query details to what your query is supposed to look like according to the msdn documentation (note that the documentation uses a `Authorization` header: sharepoint-ruby doesn't, it uses cookie attributes. Appart from that, everything should match the documentation).

There's a create exemple query on this documentation page for Sharepoint 2013.
https://msdn.microsoft.com/en-us/library/office/dn292553.aspx

**Comment by vasko** [comment]
The error message seems to come from the lack of an `X-RequestDigest` header, which is retrieved from the `/_api/contextinfo` endpoint.

I did manage to get the below code to work using direct rest calls:

```
url: http://site url/_api/web/folders
method: POST
body: { '__metadata': { 'type': 'SP.Folder' }, 'ServerRelativeUrl': '/document library relative url/folder name'}
Headers: 
    Authorization: "Bearer " + accessToken
    X-RequestDigest: form digest value
    accept: "application/json;odata=verbose"
    content-type: "application/json;odata=verbose"
    content-length:length of post body
```

I got quite a few errors while trying to make it work.
1. I had to set the Content Length header to 0 when requesting contextinfo.
2. I had to set the Content Type header when requesting contextinfo.

I don't have time to debug the gem at the moment, but when/if I do, I'll update this comment.

**Comment by piloos** [comment]
I managed to solve this issue for my usage by indeed changing the header where the token belongs to.

I change line 77 of `sharepoint-ruby.rb`

```
curl.headers["Authorization"] = "Bearer " + form_digest unless @getting_form_digest == true
```

I noticed that the python package `sharepy` is also doing it that way.

**Comment by zucaritask** [comment]
I second the above fix did work in my case for uploading files. @Plaristote do you see yourself merging https://github.com/VentureSpirit/sharepoint-ruby/commit/df5b5d6e36091c1519211f0f3a49de1292567398 any time soon?
Thanks

**Comment by Plaristote** [comment]
Well, look at that, I'm just about two years late to close that issue !
I just merged PR #33 which should fix the issue. Thank you all for the update !
